src/a11y-dialog.ts

```diff
@@ -212,17 +212,82 @@ function moveFocusToDialog(node: HTMLElement) {
   focused.focus()
 }
 
+// Elements with these ARIA roles make their children
+// `presentational`, which nullifies their semantics.
+// @see: https://www.w3.org/TR/wai-aria/
+const PRESENTATIONAL_CHILDREN_SELECTOR = `
+  a[href],
+  button,
+  img,
+  summary,
+  [role="button"],
+  [role="image"],
+  [role="link"],
+  [role="math"],
+  [role="progressbar"],
+  [role="scrollbar"],
+  [role="slider"]
+`
+
 /**
- * Get the focusable children of the given element.
+ * Get focusable children by recursively traversing the subtree of `node`.
+ * This traversal allows us to account for Shadow DOM subtrees.
  */
-function getFocusableChildren(node: HTMLElement): HTMLElement[] {
-  return $$(focusableSelectors.join(','), node).filter(
-    child =>
-      !!(
-        child.offsetWidth ||
-        child.offsetHeight ||
-        child.getClientRects().length
-      )
+function getFocusableChildren(node: ParentNode): HTMLElement[] {
+  // Check for the bases case of our recursion:
+  if (node instanceof HTMLElement) {
+    // If this node is marked as inert, neither it nor any member of
+    // its subtree will be focusable.
+    if (node.inert === true) return []
+
+    // If this node has no children, we can stop traversing.
+    // If this node has children, but nullifies its children's
+    // semantics, we can stop traversing.
+    if (
+      !node.firstElementChild ||
+      node.matches(PRESENTATIONAL_CHILDREN_SELECTOR)
+    ) {
+      // Check if the node is focusable, and then return early.
+      return isFocusable(node) ? [node] : []
+    }
+  }
+
+  let focusableEls: HTMLElement[] = []
+
+  // Walk all the immediate children of this node
+  // (with some type casting because node.children is an HTMLCollection)
+  for (const curr of node.children as unknown as HTMLElement[]) {
+    // If this element has a Shadow DOM attached,
+    // check the shadow subtree for focusable children.
+    if (!!curr.shadowRoot) {
+      focusableEls = [...focusableEls, ...getFocusableChildren(curr.shadowRoot)]
```

Comment:
We could also write it like this, although it’s worth noting it would fail if there are over 65535 nodes (the typical [maximum amount of arguments](https://stackoverflow.com/a/22747272) allowed by a function call):
```suggestion
      focusableEls.push(...getFocusableChildren(curr.shadowRoot))
```
If we do use `.push()`, we should also use it below. :)

Reply:
Hmm. I did [a benchmark](https://jsbench.me/23l6np5oic/1) for this. On my machine, it looks like:
- In Chrome/Edge, creating the new array with two spreads is fastest
- In Firefox and Safari, `oldNodes.push.apply(oldNodes, newNodes)` is fastest
- In all browsers, `oldNodes.push(...newNodes)` is slowest.

In Chromium, `push.apply` is second-fastest so I think it's the optimal choice between all 3 major browser engines.

Reply:
Amending my earlier commentary: I don't think this implementation detail will have that much impact on our users. The time it takes to do this operation is probably negligible in the real world, regardless of implementation. I'm happy to male the change we discussed, if you feel strongly about it!

Reply:
Definitely marginal if anything. :)